app/src/main/java/org/mozilla/fenix/tabstray/browser/BrowserTabsAdapter.kt

```diff
@@ -0,0 +1,87 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.tabstray.browser
+
+import android.content.Context
+import android.view.LayoutInflater
+import android.view.ViewGroup
+import androidx.recyclerview.widget.RecyclerView
+import androidx.recyclerview.widget.GridLayoutManager
+import kotlinx.android.synthetic.main.tab_tray_item.view.*
+import mozilla.components.browser.state.selector.findTab
+import mozilla.components.browser.tabstray.TabViewHolder
+import mozilla.components.browser.thumbnails.loader.ThumbnailLoader
+import mozilla.components.concept.tabstray.TabsTray
+import mozilla.components.support.base.observer.Observable
+import mozilla.components.support.base.observer.ObserverRegistry
+import org.mozilla.fenix.R
+import org.mozilla.fenix.ext.components
+import org.mozilla.fenix.ext.settings
+import org.mozilla.fenix.tabtray.TabTrayViewHolder
+
+/**
+ * A [RecyclerView.Adapter] for browser tabs.
+ */
+class BrowserTabsAdapter(
+    private val context: Context,
+    private val interactor: BrowserTrayInteractor,
+    private val layoutManager: (() -> GridLayoutManager)? = null,
+    delegate: Observable<TabsTray.Observer> = ObserverRegistry()
+) : TabsAdapter(delegate) {
+
+    /**
+     * The layout types for the tabs.
+     */
+    enum class ViewType {
+        LIST,
+        GRID
+    }
+
+    private val imageLoader = ThumbnailLoader(context.components.core.thumbnailStorage)
+
+    override fun getItemViewType(position: Int): Int {
+        return if (context.settings().gridTabView) {
+            // ViewType.GRID.ordinal
+            R.layout.tab_tray_grid_item
+        } else {
+            // ViewType.LIST.ordinal
+            R.layout.tab_tray_item
+        }
+    }
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): TabViewHolder {
+        // TODO make this into separate view holders for each layout
+        // For this, we need to separate the TabTrayViewHolder as well.
+        // See https://github.com/mozilla-mobile/fenix/issues/18535
+        val view = LayoutInflater.from(parent.context).inflate(viewType, parent, false)
+
+        return TabTrayViewHolder(view, imageLoader)
+    }
+
+    override fun onBindViewHolder(holder: TabViewHolder, position: Int) {
+        super.onBindViewHolder(holder, position)
+
+        holder.tab?.let { tab ->
+            val tabIsPrivate =
+                context.components.core.store.state.findTab(tab.id)?.content?.private == true
```

Comment:
What's the benefit of doing the '== true' as opposed to just the boolean value of .private?

Reply:
Because `findTab` is nullable, the result from `.private` is also nullable. That means we need to check that the value we want to check is not null and then test it i.e. `.private != null && .private`.

This is way is just shorthand for that expression, and also a default lint suggestion from the IDE.

Reply:
We can get rid of this entire line after we land https://github.com/mozilla-mobile/android-components/pull/9934.

Reply:
The AC PR landed, so I can update this line now to use `tab.private` instead.